pkg/diagnostics/networkpod/service.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// Description is part of the Diagnostic interface and just returns the diagnostic description.
+func (d CheckServiceNetwork) Description() string {
+	return "Check pod to service communication in the cluster. In case of ovs-subnet network plugin, all pods should be able to communicate with all services and in case of multitenant network plugin, services in non-global projects should be isolated and pods in global projects should be able to access any service in the cluster."
+}
+
+// CanRun is part of the Diagnostic interface; it determines if the conditions are right to run this diagnostic.
+func (d CheckServiceNetwork) CanRun() (bool, error) {
+	if d.KubeClient == nil {
+		return false, errors.New("must have kube client")
+	} else if d.OSClient == nil {
+		return false, errors.New("must have openshift client")
+	}
+	return true, nil
+}
+
+// Check is part of the Diagnostic interface; it runs the actual diagnostic logic
+func (d CheckServiceNetwork) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult(CheckServiceNetworkName)
+
+	pluginName, ok, err := util.GetOpenShiftNetworkPlugin(d.OSClient)
+	if err != nil {
+		r.Error("DSvcNet1001", err, fmt.Sprintf("Checking network plugin failed. Error: %s", err))
+		return r
+	}
+	if !ok {
+		r.Warn("DSvcNet1002", nil, fmt.Sprintf("Skipping service connectivity test. Reason: Not using openshift network plugin."))
+		return r
+	}
+
+	services, err := getAllServices(d.KubeClient)
+	if err != nil {
+		r.Error("DSvcNet1003", err, fmt.Sprintf("Getting all services failed. Error: %s", err))
+		return r
+	}
+	if len(services) == 0 {
+		r.Warn("DSvcNet1004", nil, fmt.Sprintf("Skipping service connectivity test. Reason: No services found."))
+		return r
+	}
+
+	localPods, _, err := util.GetLocalAndNonLocalPods(d.KubeClient)
+	if err != nil {
+		r.Error("DSvcNet1005", err, fmt.Sprintf("Getting local and nonlocal pods failed. Error: %s", err))
+		return r
+	}
+
+	var vnidMap map[string]uint32
+	if sdnplugin.IsOpenShiftMultitenantNetworkPlugin(pluginName) {
+		netnsList, err := d.OSClient.NetNamespaces().List(kapi.ListOptions{})
+		if err != nil {
+			r.Error("DSvcNet1006", err, fmt.Sprintf("Getting all network namespaces failed. Error: %s", err))
+			return r
+		}
+
+		vnidMap = map[string]uint32{}
+		for _, netns := range netnsList.Items {
+			vnidMap[netns.NetName] = netns.NetID
+		}
+	}
+
+	localGlobalPods, localNonGlobalPods := getGlobalAndNonGlobalPods(localPods, vnidMap)
+
+	// Applicable to flat and multitenant networks
+	if len(localGlobalPods) > 0 {
+		checkPodToServiceConnection(localGlobalPods[0], services[0], vnidMap, r)
+	} else {
+		r.Warn("DSvcNet1007", nil, fmt.Sprintf("Skipping service connectivity test for global projects. Reason: Couldn't find a global pod."))
+	}
+
+	// Applicable to multitenant network
+	isMultitenant := (vnidMap != nil)
+	if isMultitenant {
+		if len(localNonGlobalPods) > 0 {
+			checkPodToServiceConnection(localNonGlobalPods[0], services[0], vnidMap, r)
```

Comment:
Same issues as in the pod-to-pod tests about pod/service selection wrt VNIDs.

Also, it would be good to test both a local endpoint and a remote endpoint, if possible